// kgrabber-types/PluginContext.js
/**
 * @typedef {import("kgrabber-types/Action")} Action
 * @typedef {import("kgrabber-types/Exporter")} Exporter
 * @typedef {import("kgrabber-types/Site")} Site
 */

module.exports = class PluginContext {
	/**
	 * @param {Object} obj function for adding new actions etc.
	 * @param {function(Action):void} obj.addActionFunc
	 * @param {function(Site):void} obj.addSiteFunc
	 * @param {function(Exporter):void} obj.addExporterFunc
	 * @param {function(Step):void} obj.addStepsFunc
	 * @param {Object} obj.ui
	 * @param {Object} obj.preferences
	 * @param {Object} obj.statusManager
	 */
	constructor({ addActionFunc, addSiteFunc, addExporterFunc, addStepsFunc, ui, preferences, statusManager }) {
		this._addActionFunc = addActionFunc;
		this._addSiteFunc = addSiteFunc;
		this._addExporterFunc = addExporterFunc;
		this._addStepsFunc = addStepsFunc;
		this.ui = ui;
		this.preferences = preferences;
		this.statusManager = statusManager;
		Object.freeze(this);
	}

	/**
	 * Adds a new Action to the script
	 * @param {Action} action
	 */
	addAction(action) {
		this._addActionFunc(action);
	}

	/**
	 * Adds a new Site to the script
	 * @param {Site} site
	 */
	addSite(site) {
		this._addSiteFunc(site);
	}

	/**
	 * Adds a new Exporter to the script
	 * @param {Exporter} exporter
	 */
	addExporter(exporter) {
		this._addExporterFunc(exporter);
	}

	/**
	 * Adds a new Step to the script
	 * @param {Step} step
	 */
	addSteps(step) {
		this._addStepsFunc(step);
	}
};
