// kgrabber-plugin/Plugin.js
/**
 * @typedef {import("./PluginContext")} PluginContext
 * @typedef {import("./statusManager")} statusManager
 * @typedef {import("kgrabber-types/Action")} Action
 * @typedef {import("kgrabber-types/Exporter")} Exporter
 * @typedef {import("kgrabber-types/Site")} Site
 */

module.exports = class Plugin {
	/**
	 * @param {function(function():PluginContext):void} initFunc
	 */
	constructor(initFunc) {
		/** @type {PluginContext} */
		this.pluginContext;
		this._initFunc = initFunc;
	}

	/**
	 * @param {function():void} callback
	 */
	init(callback) {
		this.pluginContext = this._initFunc((context) => {
			this.pluginContext = context;
			callback();
		});
	}

	get ui() {
		return this.pluginContext.ui;
	}

	get preferences() {
		return this.pluginContext.preferences;
	}

	/** @type {statusManager} */
	get statusManager() {
		return this.pluginContext.statusManager;
	}

	/** @param {...Action} actions */
	addActions(...actions) {
		this.pluginContext.addActions(...actions);
	}

	/** @param {...Exporter} exporters */
	addExporters(...exporters) {
		this.pluginContext.addExporters(...exporters);
	}

	/** @param {...Site} sites */
	addSites(...sites) {
		this.pluginContext.addSites(...sites);
	}

	/** @param {Object} steps */
	addSteps(steps) {
		this.pluginContext.addSteps(steps);
	}
};
